File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.py
from collections import deque

def solution(n, edge):
    graph = [[] for _ in range(n+1)] # 1부터 n까지
    
    for a, b in edge:
        graph[a].append(b)
        graph[b].append(a)
    
    # 각각 1로부터 얼만큼 떨어져 있는지 확인 필요
    visited = [False] * (n+1)
    queue = deque([1]) # 1에서 시작
    visited[1] = True
    distance = [0] * (n+1)
    while queue:
        curr = queue.popleft()
        for neighbor in graph[curr]:
            if not visited[neighbor]:
                queue.append(neighbor)
                visited[neighbor] = True
                distance[neighbor] = distance[curr] + 1 # 부모보다 1 증가

    max_value = max(distance)
    answer = distance.count(max_value)
    return answer

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] edge) {
        List<List<Integer>> graph = new ArrayList<>();
        for(int i = 0; i <= n; i++){
            graph.add(new ArrayList<>());
        }
        
        for(int[] node : edge){
            int a = node[0];
            int b = node[1];
            graph.get(a).add(b);
            graph.get(b).add(a);
        }
        
        Queue<Integer> queue = new LinkedList<>();
        int[] distance = new int[n+1];
        
        queue.offer(1); // 1번 노드부터 시작 & depth = 0
        distance[1] = 1;
        while(!queue.isEmpty()){
            int cNode = queue.poll();
           
            for(int next : graph.get(cNode)){
                if(distance[next] == 0){
                    distance[next] = distance[cNode] + 1;
                    queue.offer(next);
                }
            }
        }
 
        int maxDist = Arrays.stream(distance).max().getAsInt();
        
        int answer = 0;
        for(int dist : distance){
            if(dist == maxDist) answer++;
        }
        return answer;
    }
}